김규화 6차 토이 리뷰 제출 : 이진탐색-매개변수탐색

rbdkqk opened the pull request:
- 10_rotatedArraySearch
- 14_binarySearchArray

rbdkqk committed:
김규화 32-insertionSort / 33-characterFrequency 제출

rbdkqk committed:
Merge : from 'upstream/master' into kim

rbdkqk committed:
Merge pull request #1 from rbdkqk/kim

merge

rbdkqk committed:
김규화 6차 토이 리뷰 제출 : 이진탐색-매개변수탐색

- 10_rotatedArraySearch
- 14_binarySearchArray

Gracechung-sw commented on 6차_이진탐색-매개변수탐색/14_binarySearchArray.js:
```diff
@@ -0,0 +1,36 @@
+  function recursion(first, last) {
+    let middleIndex = Math.floor((first + last) / 2);
+
+    if (target === array[middleIndex]) {
```
값을 찾지 못했을 때, null을 반환하고 재귀를 빠져나오는 부분은 어디서 구현되나요?

rbdkqk replied:
그렇네요! 

binarySearch([1, 2, 3, 4, 5], 6);  => Maximum call stack size exceeded
binarySearch([1, 2, 3, 4, 5], 0);  => undefined

null 처리를 해 줄 부분이 필요하겠습니다. 지적 감사합니다!

hyeonjeongchung reviewed (commented):
규화님 코드 확인하고, 궁금한 점 코멘트 달았습니다! 감사합니다.

rbdkqk commented on 6차_이진탐색-매개변수탐색/10_rotatedArraySearch.js:
```diff
@@ -0,0 +1,121 @@
+      return midIndex; // midIndex를 리턴하고 종료
+    }
+
+    if (rotated[midIndex - 2] === undefined) {
```
가령, 처리를 하다 보니 아래와 같이 진행된 경우입니다. (앞쪽 배열로 파고 들어가겠습니다.)
1회차 : [ 2,3,4,5,6,7,8] / midIndex : 3
2회차 : [ 2,3,4 ] / midIndex : 1

이 시점에서, midindex 앞에는 2(index는 0)인 값 하나만 존재합니다.

기저조건으로 target === rotated[midIndex] 만을 판단하고 있으므로, 
만약 target이 3이라면 기저조건에서 함수를 끝낼 수 있지만, 
target이 2라면 binarySearchRecursion(0, 1 - 1) 이라는 재귀를 다시 돌려야 할 것입니다.
결국 다음번 재귀에서 midindex가 0일 것이므로 기저조건에 해당하고 답이 출력될 것 같은데, 
위 지적해 주신 부분 덕분에 이 재귀 한 번을 덜 들어갈 수 있지 않을까 합니다.

'기준점 앞에 원소가 단 하나밖에 없다면' 이 하나의 원소와 target만 비교하면 되기 때문입니다.

그런데 생각해 보니 기준점의 앞에서만 이런 처리를 해 주고 있고, 기준점의 뒤에 하나만 남았을 경우는 작성하지 않았네요..

hyunju-song commented on 6차_이진탐색-매개변수탐색/10_rotatedArraySearch.js:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  return binarySearchRecursion(firstIndex, lastIndex);
+  // 전에 짠 코드는 불완전했음. https://github.com/codestates/help-desk/issues/1628
```
만약에 주어진 배열이 [7 ,8, 9,10, 0, 1 , 2, 3, 4, 5, 6 ] 이고,  찾는 값이 10인 경우에는 위의 코드에서 null로 나올수도 있지 않나요?  꼭 정렬된 배열의 중간을 기준으로 회전시켰다고는 하지 않아서 해당 부분에 대해서 질문드립니다!

hyunju-song replied:
아하 그렇군요!! 사실 규화님이랑 현정님 코드로도 테스트는 통과되는듯 한데, 왜 되는지가 제가 잘 이해가 안가서 질문 남겼습니다! 감사합니다!

rbdkqk committed:
fix : toy-14

null 처리가 누락된 점을 보완함

rbdkqk commented on 6차_이진탐색-매개변수탐색/10_rotatedArraySearch.js:
```diff
@@ -0,0 +1,121 @@
+
+    if (rotated[first] < rotated[midIndex - 1]) {
+      // 기준점의 앞쪽이 제대로 정렬되어 있다면,
+      if (target >= rotated[first] && target < rotated[midIndex]) {
```
이 조건은 잘못되었습니다.

```
 if (target >= rotated[first] && target < rotated[midIndex])

=> 

 if (target >= rotated[first] && target <= rotated[midIndex -1 ])
```

이렇게 해 줘야 정상 동작하는 것 같습니다.

rbdkqk committed:
fix : toy-10

42번째 줄의 if문 이하는 없어도 동작한다.
  - 재귀를 한번 덜 들어가게 하는 정도의 효과인 듯?

54번째 줄의 조건 수정함 (테스트 결과 정상 작동하지 않았었음)
  - 기존 : if (target >= rotated[first] && target < rotated[midIndex])
  - 수정 : if (target >= rotated[first] && target <= rotated[midIndex - 1])

rbdkqk commented on 6차_이진탐색-매개변수탐색/10_rotatedArraySearch.js:
```diff
@@ -0,0 +1,122 @@
+
+    if (rotated[first] < rotated[midIndex - 1]) {
+      // 기준점의 앞쪽이 제대로 정렬되어 있다면,
+      if (target >= rotated[first] && target <= rotated[midIndex - 1]) {
```
이 부분은 잘못되었었습니다.

if (target >= rotated[first] && target < rotated[midIndex ])  이런 상태였는데,

if (target >= rotated[first] && target <= rotated[midIndex - 1]) 이렇게 수정하였습니다.

rbdkqk commented on 6차_이진탐색-매개변수탐색/14_binarySearchArray.js:
```diff
@@ -0,0 +1,41 @@
+      return middleIndex; // index를 리턴한다.
+    }
+
+    // 탐색을 마쳤는데 끝까지 결과값이 안 나왔다면,
```
바로 위에서 지적해 주신 null 리턴을 위해 이 코드를 추가했습니다.

rbdkqk commented on 6차_이진탐색-매개변수탐색/10_rotatedArraySearch.js:
```diff
@@ -0,0 +1,122 @@
+      }
+    } else {
+      // 그 외에는, (기준점의 뒤쪽이 제대로 정렬되어 있다면,)
+      if (target >= rotated[midIndex + 1] && target <= rotated[last]) {
```
`if (target >= rotated[midIndex + 1] && target <= rotated[last]) ` 이 조건 중 `target <= rotated[last]` 부분에 관하여,

앞선 리뷰 시간에 last 값이 68번째 줄에 의해 음수가 된다면 이 줄의 `rotated[last]` 은 `rotated[-1]`이 되는데, 

인덱스가 -1인 경우 rotated[last]는 `undefined`가 되고,  target(일정한 숫자)과 `undefined`를 비교하는 연산을 수행하는 점에 관하여 대화가 있었습니다.


즉, 가령 target이 2라면, 위 if 조건문의 우측 부분은 `2 <= undefined`가 true인지 false인지에 따라 분기를 나누게 됩니다.

---

그러나, 이는 자바스크립트의 약점을 이용한 것이므로 좋지 않다는 지적을 받았습니다. 

숫자와 undefined를 비교 연산할 수는 없는 것이므로, 이 부분에 관한 처리를 별도로 해 주어야 합니다.

그래서 재귀를 위한 52번째 줄 앞에서, last가 0보다 작을 경우 곧바로 null을 리턴하는 처리를 추가해야 할 것으로 보입니다.

현정 님께서 작성하셨던, 이 부분에 관한 처리를 해 주신 코드가 정확한 코드라고 해야 하겠습니다.

앞선 대화에서 "그냥 잘 동작하니까 넘어갔던" 사항을, 제가 따로 확인하여 위와 같이 지적받았기에, 이를 수정하고 공유해 드립니다!
